### ISSUE TITLE
[Question] MultiTaskGP for vector valued function with correlated outputs

### ISSUE THREAD

**Issue by fleskovar** [pytorch/botorch#736]
I am looking into implementing a model that produces multiple correlated output for multiple tasks (if I am not mistaken this should be what the documentation describes as multi-task-multi-output - MTMO). For this type of model, I would assume that the input tensor has the shape _n x d+1_ (d inputs plus the additional task index) while the output tensor has the shape _n x o_ (where o is the number of correlated outputs).

Even though the documentation says that MultiTaskGP class should allow me to implement what I am looing for, the expected shape of the training observations is _n_ given training data _n x d+1_.

How would I go about implementing this type of model? Is MultiTaskGP the right type of model for this?

**Comment by Balandat** [comment]
Interesting. So effectively you could reduce your model to the one in MultiTaskGP by introducing additional task indices: Say you have 2 tasks with indices `{0, 1}`, and two outputs for each task. Then you could have a standard Multi-Task model `{00, 01, 10, 11}` where `00` corresponds to the first output of the first task, `01` corresponds to the second output of the first task etc. 

Whether that's exactly what you'd like to do is not clear and depends on your situation. Presumably just using a full task covariance matrix on the resulting model might overparameterize the model if you believe that the cross-outcome correlation is the same (or similar) for the different tasks. In that case it would make sense to re-parameterize this using a structured covariance. 

In your case, are all outcomes observed for all tasks? I.e. if you have `n_t` tasks and `n_o` outcomes, will you always observe a `n_t*n_o` values together? In that case your model could be set up efficiently using Kronecker structure in the cross-outcome covariance matrix. Structurally, such model would be a very similar to the High Order Gaussian Process model of https://arxiv.org/pdf/1911.06287.pdf

If you only have partial observations the same idea about constructing the covariance matrix applies, but it won't have nice Kronecker structure anymore. 

Does this make sense?

cc @qingfeng10, @wjmaddox

**Comment by fleskovar** [comment]
Thank you so much for the quick reply. Apologies in advance if the ideas/question do not make too much sense. I am still learning a lot about GPs and BO.

For this particular model all outcomes are observed simultaneously but not all tasks are. Similarly to what is described in the paper below, I would like to use results from the optimization of a given problem to avoid doing a cold start on a related/similar problem:

https://papers.nips.cc/paper/2013/file/f33ba15effa5c10e873bf3842afb46a6-Paper.pdf

Indeed, cross-outcome correlation between tasks should be similar, so the full task covariance matrix might not be suitable. Do you think that dividing the covariance matrix into three components might be a good idea?

k([x, i, l],[x′, j, m])=k_inputs(x,x′)∗k_outcomes(i,j)∗k_task(l, m)

With this setup k_inputs∗k_outcomes would be computed with a Kronecker product while the task kernel is introduced as a Hadamard product. Do you think this makes sense?

**Comment by wjmaddox** [comment]
Yes, I think that makes sense. Here's a really basic (untested) model class for what you're defining:

```python
from torch import Tensor
from gpytorch.kernels import MaternKernel, IndexKernel, MultitaskKernel
from gpytorch.means import MultitaskMean, ConstantMean
from gpytorch.lazy import KroneckerProductLazyTensor
from gpytorch.likelihoods import MultitaskGaussianLikelihood
from gpytorch.models import ExactGP
from botorch.models.gpytorch import GPyTorchModel
from gpytorch.distributions import MultitaskMultivariateNormal

class MultiOutputMultiTaskGP(ExactGP, GPyTorchModel):
    r"""
    This multi task GP is more like the multi task gp in 
    https://github.com/cornellius-gp/gpytorch/blob/master/examples/03_Multitask_Exact_GPs/Multitask_GP_Regression.ipynb
    rather than the multi task GP implemented in botorch.models.MultiTaskGP, due to not
    having explicit task indices.
    """
    def __init__(
        self,
        train_X: Tensor,
        train_Y: Tensor,
        likelihood: MultitaskGaussianLikelihood = None,
        rank = None,
        lik_rank = None,
    ) -> None:

        num_tasks = train_Y.shape[-1]
        batch_shape, ard_num_dims = train_X.shape[:-2], train_X.shape[-1]
        
        if lik_rank is None:
            lik_rank = rank
                       
        self._validate_tensor_args(X=train_X, Y=train_Y)

        if likelihood is None:
            likelihood = MultitaskGaussianLikelihood(
                num_tasks=num_tasks, 
                rank=lik_rank if lik_rank is not None else 0,
            )

        super(MultiOutputMultiTaskGP, self).__init__(train_X, train_Y, likelihood)
        self._rank = rank if rank is not None else num_tasks
        self._num_outputs = num_tasks

        self.mean_module = MultitaskMean(
            ConstantMean(), num_tasks=num_tasks
        )
        
        self.data_kernel = MaternKernel()
        self.task_kernel = IndexKernel(num_tasks=len(torch.unique(train_X[..., -1])))
        self.output_kernel = IndexKernel(num_tasks=num_tasks)
        
        self.to(train_X)
        
    def forward(self, x: Tensor) -> MultitaskMultivariateNormal:
        mean_x = self.mean_module(x)
        data_and_task_x = self.data_kernel(x[..., :-1]).mul(self.task_kernel(x[..., -1].long()))
        output_x = self.output_kernel.covar_matrix
        covar_x = KroneckerProductLazyTensor(data_and_task_x, output_x)
        return MultitaskMultivariateNormal(mean_x, covar_x)

# first index is the data, second index is the task indicator
train_x = torch.cat((torch.linspace(0, 1, 100).unsqueeze(-1), (torch.rand(100,1)>0.5).float()),dim=1)

# generate a multitask response that is n x o (here o = 5)
train_y = torch.sin(2. * train_x[:,0] * train_x[:,1].unsqueeze(0) * torch.arange(3, 8).unsqueeze(1)).t()

# define model
model = MultiOutputMultiTaskGP(train_x, train_y)

# sample from prior to demonstrate shapes
model(train_x).rsample(torch.Size((32,))).shape # should be 32 x 100 x 5
```

@Balandat  we should probably polish up #637 so that this type of model is more straightforward to implement in the future.

**Comment by Balandat** [comment]
I finally found time to clean up #637, which should do what you want.

**Comment by fleskovar** [comment]
Apologies the late reply. I have been looking at the `KroneckerMultiTaskGP` at it seems to applicable only for the multi-output scenario where the kernel is calculated as follows:
![image](https://user-images.githubusercontent.com/6884660/133938711-6c961312-b01f-443f-8f25-ad7051514fa8.png) (1)

In the model implemented above by @wjmaddox the kernel has the following form:
![image](https://user-images.githubusercontent.com/6884660/133938774-140a69ec-d6ba-41ec-921a-0dc5bda302c8.png) (2)

I though about leveraging the implementation of the `MultitaskKernel` and adding the kernel for the different tasks as follows inside the `forward` method of the model:
![image](https://user-images.githubusercontent.com/6884660/133938876-a53f38b6-b478-4ac3-b63d-8a5cd04ac688.png) (3)
where:
![image](https://user-images.githubusercontent.com/6884660/133939005-eb288b2f-631e-4bec-bccf-9a066f51618e.png) (4)
I am not sure if this would be the recommended approach or if I should try to implement a new Kernel class and pass it as the `data_covar_module` argument of the  `MultitaskKernel` . Ideally I would like to take advantage of the implementation of the `posterior` method of the `MultitaskKernel`.

Addtionally, I have also been looking at the `posterior` function of the `KroneckerMultiTaskGP`, specifically the code below:
```python
# build a root decomposition of the joint train/test covariance matrix
# construct (\tilde{R} \otimes M)(\tilde{R} \otimes M)' \approx
# (K_{(x,xt), (x,xt)} \otimes Ktt)
joint_covar = RootLazyTensor(
    KroneckerProductLazyTensor(updated_root, task_covar_rootlt.root.detach())
 )
```
Is this correct or should the terms `updated_root` and `task_covar_rootlt.root.detach()` be switched to get an expression similar to (1)? My understanting is that the following expression usually holds for kronecker products:
![image](https://user-images.githubusercontent.com/6884660/133939391-cfade98b-2c90-4ba8-9b28-2283e0da08d6.png)

**Comment by wjmaddox** [comment]
I'd have to double check that but in general the kernel should be `K_{xx} \otimes K_{tt}` rather than `K_{tt} \otimes K_{xx}`. This is generically self-consistent with the implementation in gpytorch proper [here](https://github.com/cornellius-gp/gpytorch/blob/f06004e888542289d19c155e4e2ad35aad0573b7/gpytorch/kernels/multitask_kernel.py#L41).
